Source/Core/DolphinQt2/Translation.cpp

```diff
@@ -0,0 +1,220 @@
+      return;
+    }
+
+    u32 magic = ReadU32(&m_data[0]);
+    if (magic != MO_MAGIC_NUMBER)
+    {
+      ERROR_LOG(COMMON, "MO file '%s' has bad magic number %x\n", filename.c_str(), magic);
+      m_data = {};
+      return;
+    }
+
+    u16 version_major = ReadU16(&m_data[4]);
+    if (version_major > 1)
+    {
+      ERROR_LOG(COMMON, "MO file '%s' has unsupported version number %i", filename.c_str(),
+                version_major);
+      m_data = {};
+      return;
+    }
+
+    m_number_of_strings = ReadU32(&m_data[8]);
+    m_offset_original_table = ReadU32(&m_data[12]);
+    m_offset_translation_table = ReadU32(&m_data[16]);
+  }
+
+  u32 GetNumberOfStrings() const { return m_number_of_strings; }
+  const char* Translate(const char* original_string) const
+  {
+    auto iter = std::lower_bound(
+        MoIterator(m_data.data(), m_offset_original_table),
+        MoIterator(m_data.data(), m_offset_original_table, m_number_of_strings), original_string,
+        [](const char* a, const char* b) { return strcmp(a, b) < 0; });
+
+    if (strcmp(*iter, original_string) != 0)
+      return original_string;
+
+    u32 offset = ReadU32(&m_data[m_offset_translation_table + iter.GetIndex() * 8 + 4]);
+    return &m_data[offset];
+  }
+
+private:
+  std::vector<char> m_data;
+  u32 m_number_of_strings = 0;
+  u32 m_offset_original_table = 0;
+  u32 m_offset_translation_table = 0;
+};
+
+class MoTranslator : public QTranslator
+{
+public:
+  using QTranslator::QTranslator;
+
+  bool load(const std::string& filename)
+  {
+    m_mo_file = MoFile(filename);
+    return m_mo_file.GetNumberOfStrings() > 0;
+  }
+
+  QString translate(const char* context, const char* source_text,
+                    const char* disambiguation = Q_NULLPTR, int n = -1) const override
+  {
+    return QString::fromUtf8(m_mo_file.Translate(source_text));
+  }
+
+private:
+  MoFile m_mo_file;
+};
+
+static bool TryInstallExactTranslator(const std::string& language)
+{
+  auto* translator = new MoTranslator(QApplication::instance());
+  auto filename =
+#if defined _WIN32
+      File::GetExeDirectory() + StringFromFormat("/Languages/%s/dolphin-emu.mo", language.c_str())
+#elif defined __APPLE__
+      File::GetBundleDirectory() +
+      StringFromFormat("/Contents/Resources/%s.lproj/dolphin-emu.mo", language.c_str())
+#else
+      StringFromFormat(DATA_DIR "/../locale/%s/LC_MESSAGES/dolphin-emu.mo", language.c_str())
+#endif
+      ;
+  if (translator->load(filename))
+  {
+    QApplication::instance()->installTranslator(translator);
+    return true;
+  }
+  else
+  {
+    translator->deleteLater();
+    return false;
+  }
+}
+
+static bool TryInstallTranslator(const std::string& language)
+{
+  bool result = true;
+  auto underscore_position = language.find('_');
+  if (underscore_position != std::string::npos)
+    result &= TryInstallExactTranslator(language.substr(0, underscore_position));
+  return result && TryInstallExactTranslator(language);
```

Comment:
If there is an underscore, the return value should be true if either variant was found. Otherwise this isn't going to work for languages like Chinese (which we have `zh_CN` and `zh_TW` files for but no `zh` file).